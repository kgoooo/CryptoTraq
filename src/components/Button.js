/*
		This is the new button component that I made as a result of adding Radium.  Here I can keep my code 'DRY'
		by having just one Button component.  I wanted to remove CSS files completely by using Radium for this app
		and found it to be impossible to not repeat code unless the app is truly reusable with its components.
 */

import React from 'react';
import PropTypes from 'prop-types';
import Radium from 'radium';
import RadiumVars from '../RadiumVariables';

/*  Stateless Button components, the button content and methods get passed through props now.  Before each
		button was configured manually.
*/
/*
		Added the animation style for the button, passed in through props.
 */
const Button = (props) => {
	return (
		<button
			onClick={props.onClick}
			style={[styles.btn__main, props.buttonAnimation]}
		>{props.buttonContent}</button>
	)
};

Button.propTypes = {
	onClick: PropTypes.func.isRequired,
	buttonContent: PropTypes.string.isRequired,
	buttonAnimation: PropTypes.object
};

/*  Button styles for Radium defined here and added below to the component on the export statement. */
const styles = {
	btn__main: {
		background: 'transparent',
		borderRadius: '8px',
		border: `2px solid ${RadiumVars.color.colorPrimaryLightBlue}`,
		color: RadiumVars.color.colorPrimaryLightBlue,
		fontSize: 'inherit',
		fontWeight: '300',
		padding: '.5rem 2.5rem',
		transition: 'all ease-in-out .3s',

		':hover': {
			boxShadow: `0 0 3rem ${RadiumVars.color.colorPrimaryBlue}`
		}
	}
};

export default Radium(Button);
