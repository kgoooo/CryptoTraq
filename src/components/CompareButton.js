import React from 'react';

const CompareButton = (props) => {
	return (
		<div  className="btn--compare">
			<button className="btn btn--light" onClick={props.exchangeSearch}>Track</button>
		</div>
	)
}

export default CompareButton;