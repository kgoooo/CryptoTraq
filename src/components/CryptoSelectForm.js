/*  I refactored this to be a sub component for the new SelectionForm component.
		Once again I realized that to keep the code DRY I needed to refactor the selection
		components to make them more reusable.  This now gets passed into Selection form
		to render the proper content for the select form elements.
		---The main reason why there is still this component and the IntlSelectForm is
		because I couldn't get the crypto.name to render conditionally while being mapped.
*/

import React from 'react';
import Radium from 'radium';
import RadiumVars from '../RadiumVariables';


const CryptoSelectForm = (props) =>{
	return (
		<select style={styles.selectForm__form}>
			{props.list.map(crypto => (
				<option key={crypto.code} value={crypto.code}>{crypto.name} â€“ {crypto.code}</option>
			))}
		</select>
	)
};

// Radium styles

const styles = {
	selectForm__form: {
		backgroundColor: RadiumVars.color.colorPrimaryDarkBlue,
		border: `2px solid ${RadiumVars.color.colorPrimaryBlue}`,
		borderRadius: '8px',
		color: RadiumVars.color.colorPrimaryLightBlue,
		fontSize: 'inherit',
		height: '32px',
		outline: 'none',
	},
};


export default Radium(CryptoSelectForm);
