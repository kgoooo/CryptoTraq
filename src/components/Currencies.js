import React from 'react';

const Currencies = (props) => {
	const cryptoList = props.cryptoList;
	const intlList = props.intlList;
	return (
		<div>
      <div>
				<ul>{cryptoList.map(c => <li key={c.code}>{c.name} - {c.code}</li>)}</ul>
        <ul>{intlList.map(i => <li key={i.code}>{i.name}</li>)}</ul>
      </div>
		</div>
	)
}

export default Currencies;