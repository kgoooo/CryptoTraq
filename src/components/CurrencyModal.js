import React, { Component } from 'react';
import ReactModal from 'react-modal';
import SupportedCurrencies from './SupportedCurrencies';

class CurrencyModal extends Component {
  constructor(props) {
		super(props);
		
		this.state = { isOpen: false };
  }
  handleOpenModal = () => {
    this.setState({ isOpen: true });
	}
	
	handleCloseModal = () => {
    this.setState({ isOpen: false });		
	}
  render() {
    let isOpen = this.props.isOpen;
    return (
      <div>
				<SupportedCurrencies modalOpen={this.handleOpenModal} />
				<ReactModal 
					isOpen={this.state.isOpen}
					contentLabel="Modal showing available currencies supported by this app."
					className="Modal"
					overlayClassName="Overlay"
				>
					<button className="btn btn--light" onClick={this.handleCloseModal}>X</button>
				</ReactModal>
      </div>
    );
  }
}

export default CurrencyModal;