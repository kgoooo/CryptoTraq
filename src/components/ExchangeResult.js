import React from 'react';

const ExchangeResult = (props) => {
	let showRate;
	if(props.rate !== 0){
		showRate = <h3 className="rate__text">The exchange rate per {props.crypto} is <span className="rate__span">{props.rate > 1 ? Math.round(props.rate) : props.rate}</span> {props.intl}</h3>
	}
	if(props.excError === true) {
		console.log("banans");
		showRate = <h3 className="rate__text">The exchange rate could not be calculated for these currencies, please try another option.</h3>
	}
	return(
		<div className="rate">
			{showRate}
		</div>
	)
}

export default ExchangeResult;