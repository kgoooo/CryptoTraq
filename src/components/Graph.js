/*
		I added Radium to this component as well.  Small amount of styling is used as the actual graph generated
		is a canvas object.  I changed some of the styling in the graph options object to make it easier
		to understand/read for the user.  I originally did not optimize the coloring of the graph information,
		and now its an overall better experience.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types'
import Radium from 'radium';
import {Line} from 'react-chartjs-2';
import moment from 'moment';
import RadiumVars from '../RadiumVariables';

class Graph extends Component {
	constructor(props) {
		super(props);
	}

	render(){
		let zuluArray = [];
		let dataArray = [];

		this.props.cryptoData.forEach(el => {
			zuluArray.push(moment(el.zuluTime).format("DD-MMM HH:mm"));
			dataArray.push(el.rate)
		});
		const data = {
			labels: zuluArray,
			datasets: [{
				label: 'Currency value',
				fill: false,
				lineTension: 0.3,
				backgroundColor: 'rgba(75,192,192,0.4)',
				borderColor: '#00FFF5',
				borderCapStyle: 'butt',
				borderJoinStyle: 'round',
				pointBackgroundColor: 'rgba(255, 255, 255, .8)',
				pointBorderWidth: 1,
				pointHoverRadius: 5,
				pointHoverBackgroundColor: 'rgba(75,192,192,1)',
				pointHoverBorderColor: 'rgba(220,220,220,1)',
				pointHoverBorderWidth: 2,
				pointRadius: 1,
				pointHitRadius: 10,
				data: dataArray
			}]
		};
		return(
			<div>
				{dataArray.length > 0 ? <div style={styles.graph} >
					<Line 
					data={data} 
					width={95} 
					height={600}
					options={{
						maintainAspectRatio: false,
						title: {
							fontSize: 20,
							fontColor: '#00FFF5',
							fontFamily: "'Roboto', sans-serif",
							fontStyle: 'normal',
							display: true,
							padding: 25,
            	text: `Value of ${this.props.crypto} over the past 24 hours.`
						},
						scales: {
							yAxes: [{
								ticks: {
									fontColor: "#00ADB5",
								},
								gridLines: {
									color: "rgba(255, 255, 255, 0.1)"
								},
							}],
							xAxes: [{
								ticks: {
									fontColor: "#00ADB5",
								},
								gridLines: {
									color: "rgba(255, 255, 255, 0.1)"
								},
							}]
						},
						legend: {
							labels: {
								fontColor: '#00FFF5'
							}
						}
					}}
				/>
				</div>: null}
			</div>
		)
	}
}

Graph.propTypes = {
	cryptoData: PropTypes.array.isRequired,
	crypto: PropTypes.string
};

const styles = {
	graph: {
		width: '85%',
		border: `3px solid ${RadiumVars.color.colorPrimaryBlue}`,
		borderRadius: '8px',
		marginLeft: 'auto',
		marginRight: 'auto',
		marginBottom: '55px',

		'@media (max-width: 500px)': {
			width: '95%',
			marginBottom: '20px'
		}
	}
};

export default Radium(Graph);
export {Graph};