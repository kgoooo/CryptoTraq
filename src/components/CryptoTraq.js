/*
		Did some cleaning up of imports in this component and added Radium for the arrow
		styling.
		This component had some refactoring when I made the SelectionForm component to
		break up the big form component I had before.  Mainly adding the new components
		and changing the props that are passed in accordingly.
		Button component is also used here now instead of CompareButton.
 */
/*
	I added two loading spinner components to be rendered here.  One will show up when the app
	is initializing, especially if the user is on a slower connection.  The second will show
	when the user has tracked currencies and is waiting for the results.
 */

import React from 'react';
import PropTypes from 'prop-types';
import Radium from 'radium';
import Button from './Button';
import SelectionForm from './SelectionForm';
import CryptoSelectForm from './CryptoSelectForm';
import IntlSelectForm from './IntlSelectForm';
import ExchangeResult from './ExchangeResult';
import LoadingSpinner from './LoadingSpinner';
import Graph from './Graph';
import GraphOptionsBar from './GraphOptionsBar';
import RadiumVars from '../RadiumVariables';

const CryptoTraq = (props) => {
  return(
    <div>
        <div>
          <SelectionForm
              formLabel={"Select Cryptocurency:"}
              selectionFunc={props.handleCryptoChange}
              selectForm={<CryptoSelectForm list={props.cryptoList} delay={styles.delay1} anim={styles.select__glow}/>}
          />
          <p style={styles.arrow}>&rarr;</p>
          <SelectionForm
            formLabel={"Select international currency:"}
            selectionFunc={props.handleIntlChange}
						selectForm={<IntlSelectForm list={props.intlList} delay={styles.delay2} anim={styles.select__glow}/>}
					/>
          <Button onClick={props.handleExchangeSearch} buttonContent={"Track currency"} buttonAnimation={styles.trackAnim}/>
        </div>
				<div>
					<GraphOptionsBar handleRangeZeroChange={props.handleRangeZeroChange} rangeContent={props.rangeContent}/>
				</div>
				{props.initializing ?
					<LoadingSpinner
						loading={props.initializing}
						content={"Loading currencies"}
					/> : null}
				{props.loading ?
					<LoadingSpinner
						loading={props.loading}
						content={'Fetching your results, please hold on.'}
					/>: null}
        {props.showingExchangeRate === true ? (
          <ExchangeResult
            crypto={props.selectedCrypto}
            intl={props.selectedIntl}
            rate={props.exchangeRate}
            excError={props.showingExchangeError}
          />
        ) : null}
        <Graph
          cryptoData={props.dataSet}
          crypto={props.selectedCrypto}
					rangeZero={props.rangeZero}
        />
			</div>
  )
};

CryptoTraq.propTypes = {
	cryptoList: PropTypes.array.isRequired,
	dataSet: PropTypes.array.isRequired,
	exchangeRate: PropTypes.number.isRequired,
	initializing: PropTypes.bool.isRequired,
	intlList: PropTypes.array.isRequired,
	loading: PropTypes.bool.isRequired,
	rangeContent: PropTypes.string.isRequired,
	rangeZero: PropTypes.bool.isRequired,
	selectedCrypto: PropTypes.string,
	selectedIntl: PropTypes.string,
	showingExchangeError: PropTypes.bool.isRequired,
	showingExchangeRate: PropTypes.bool.isRequired,
	handleCryptoChange: PropTypes.func.isRequired,
	handleIntlChange: PropTypes.func.isRequired,
	handleExchangeSearch: PropTypes.func.isRequired,
};

//  Radium Styles
/*
		Added selectGlowAnim to add a small animation to the selection element to indicate to the user
		that they should take action there first.  Applied with select__glow
		I added this trackBtnAnimation to add a small scale animation to the track button as the
		3rd step to tracking your conversion rate.  It gets passed in as props and set in the button component
 */

const selectGlowAnim = Radium.keyframes({
	'50%': {boxShadow: `0 0 3px 6px ${RadiumVars.color.colorPrimaryLightBlue}`},
});
const trackBtnAnimation = Radium.keyframes({
	'0': {transform: 'scale(1)'},
	'50%': {
		transform: 'scale(1.1)',
		boxShadow: `0 0 3px 6px ${RadiumVars.color.colorPrimaryLightBlue}`
	},
	'100%': {transform: 'scale(1)'}
}, 'track');

const styles = {
	arrow: {
		display: 'inline-block',
		fontSize: '50px',
		color: RadiumVars.color.colorPrimaryBlue,
		textShadow: `0px 0px 25px ${RadiumVars.color.colorPrimaryBlue}`,

		'@media (max-width: 500px)': {
			margin: '0px'
		}
	},
	select__glow: {
		animationName: selectGlowAnim,
		animationDuration: '1.5s',
	},
	trackAnim: {
		animationName: trackBtnAnimation,
		animationDuration: '1.5s',
		animationDelay: '3.5s'
	},
	delay1: { animationDelay: '0.85s' },
	delay2: { animationDelay: '2.35s'}
};

export default Radium(CryptoTraq);