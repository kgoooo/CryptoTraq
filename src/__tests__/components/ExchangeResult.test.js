import React from 'react';
import { shallow } from 'enzyme';
import { ExchangeResult } from '../../components/ExchangeResult';

describe('<ExchangeResult />', () => {

	it('Should render correctly', () => {
		const wrapper = shallow(<ExchangeResult rate={76.34} crypto={"BTC"} intl={"USD"}/>);
		expect(wrapper).toMatchSnapshot();
	});

	it('Should render rate correctly, when rate and currencies provided', () => {
		const wrapper = shallow(<ExchangeResult rate={76.34} crypto={"BTC"} intl={"USD"}/>);
		expect(wrapper.find('h3')).toHaveLength(1);
		expect(wrapper.text()).toEqual('The exchange rate per BTC is 76 USD')
	});

	it('should show error when no exchange rate found // error state is true.', () => {
	const wrapper = shallow(<ExchangeResult crypto={"BTC"} intl={"USD"} excError={true}/>);
		expect(wrapper.text()).toEqual("The exchange rate could not be calculated for these currencies, please try another option.")
	})
});
