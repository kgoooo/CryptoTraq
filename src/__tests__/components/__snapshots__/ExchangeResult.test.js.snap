// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExchangeResult /> Should render correctly 1`] = `
<div>
  <h3
    style={
      Object {
        "@media (max-width: 500px)": Object {
          "padding": "20px",
          "textAlign": "center",
        },
        "color": "#00FFF5",
        "fontSize": "24px",
        "fontWeight": "300",
        "marginBottom": "30px",
        "textAlign": "center",
      }
    }
  >
    The exchange rate per 
    BTC
     is 
    <span
      style={
        Object {
          "fontWeight": "500",
        }
      }
    >
      76
    </span>
     
    USD
  </h3>
</div>
`;
